pkg/pulumiyaml/analyser.go

```diff
@@ -8,14 +8,255 @@ import (
+	tc.resourceNames[name] = resource
+	tc.resources[resource] = typ
+}
+
+// notAssignable represents the logic chain for why an assignment is not possible.
+type notAssignable struct {
+	reason   string
+	because  []*notAssignable
+	internal bool
+	property string
+}
+
+func (n notAssignable) String() string {
+	return n.string(0)
+}
+
+func (n notAssignable) string(indent int) string {
+	var prop string
+	if n.property != "" {
+		prop = n.property + ": "
+	}
+	s := strings.Repeat("  ", indent) + prop + n.reason
+	if len(n.because) > 0 {
+		s += ":"
+	}
+	for _, arg := range n.because {
+		s += "\n" + arg.string(indent+1)
+	}
+	return s
+}
+
+func (n notAssignable) IsInternal() bool {
+	if n.internal {
+		return true
+	}
+	for _, b := range n.because {
+		if b.IsInternal() {
+			return true
+		}
+	}
+	return false
+}
+
+func (n notAssignable) Because(b ...*notAssignable) *notAssignable {
+	n.because = b
+	return &n
+}
+
+func (n notAssignable) Property(propName string) *notAssignable {
+	n.property = propName
+	return &n
+}
+
+// isAssignable determines if the type `from` is assignable to the type `to`.
+// If the assignment is legal, nil is returned.
+func isAssignable(from, to schema.Type) *notAssignable {
+	to = codegen.UnwrapType(to)
+	from = codegen.UnwrapType(from)
+
+	// If either type is invalid, we return. An error message should have
+	// already been generated, we don't need to add another one.
+	if _, ok := from.(*schema.InvalidType); ok {
+		return nil
+	}
+	if _, ok := to.(*schema.InvalidType); ok {
+		return nil
+	}
+
+	fail := &notAssignable{
+		reason:   fmt.Sprintf("Cannot assign %s to type %s", from, to),
+		internal: false,
+	}
+	okIf := func(cond bool) *notAssignable {
+		if !cond {
+			return fail
+		}
+		return nil
+	}
+	okIfAssignable := func(s *notAssignable) *notAssignable {
+		if s == nil {
+			return nil
+		}
+		return fail.Because(s)
+	}
+
+	// If from is a union type, then it can only assign if every element it contains can
+	// be assigned to `to`.
+	if from, ok := from.(*schema.UnionType); ok {
+		reasons := []*notAssignable{}
+		for _, subtype := range from.ElementTypes {
+			because := isAssignable(subtype, to)
+			if because != nil {
+				reasons = append(reasons, because)
+			}
+		}
+		if len(reasons) == 0 {
+			// every assignment worked
+			return nil
+		}
+		// at least 1 assignment didn't work, so we fail
+		return fail.Because(reasons...)
+	}
+
+	if schema.IsPrimitiveType(to) {
+		switch to {
+		case schema.AnyType:
+			return okIf(true)
+		case schema.NumberType, schema.IntType:
+			return okIf(from == schema.NumberType || from == schema.IntType)
+		case schema.StringType:
+			// Resources can coerce into strings (by implicitly calling urn)
+			_, isResource := from.(*schema.ResourceType)
+			return okIf(isResource || from == schema.StringType ||
+				// Since we don't have a fn(number) -> string function, we coerce numbers to strings
```

Comment:
Also confused on this.

Reply:
I am attempting to preserve the current behavior. Consider the following example, taken from the `aws-fargate-pp` example. Consider this snippet:

```yaml
  ...
  appTask:
    type: aws:ecs:TaskDefinition
    properties:
      family: fargate-task-definition
      cpu: 256
      memory: 512
  ...
```

Here both [`cpu`](https://github.com/pulumi/pulumi-aws/blob/master/provider/cmd/pulumi-resource-aws/schema.json#L256979-L256983) and [`memory`](https://github.com/pulumi/pulumi-aws/blob/master/provider/cmd/pulumi-resource-aws/schema.json#L257006-L257010) take strings, but are given as numbers. This just works. I didn't want to break existing code.

Reply:
Also of note, if we image a Pulumi YAML program with an important number that is refactored to a variable and then given to properties of number and string types:
```yaml
variables:
  myNumber: 42
resources:
  handleNumber:
    asNumberType: ${myNumber}
    asStringType: ${myNumber}
```

I feel like we want this to just work. Other languages would have us call `{ asNumberType: myNumber, asStringType: myNumber.string() }` but we don't have that support in Pulumi YAML. We could require that the user calls `Fn::ToJSON`, but I don't think we want to require that.

Reply:
That explanation makes perfect sense.